File: linkedList.py
#A simpler implementation
class node:
    def __init__(self, data = None, next = None) :
        self.data = data
        self.next = next
    
class linked_list :
    def __init__(self) :
        self.head = None
    
    def addToFront(self, data) :
        self.head = node(data = data, next = self.head)

    def isEmpty(self) :
        return self.head == None

    def addAtEnd(self, data) :
        if not self.head :
            self.head = node(data = data)
            #return
        curr = self.head
        while curr.next :
            curr = curr.next
        curr.next = node(data = data)

    def deleteNode(self, key) :
        curr = self.head
        prev = None
        while curr and curr.data != key :
            prev = curr
            curr = curr.next
        if prev is None :
            self.head = curr.next
        elif curr :
            prev.next = curr.next
            curr.next = None

    def getLastNode(self) :
        temp = self.head
        while(temp.next is not None) :
            temp = temp.next
        return temp.data

    def printList(self) :
        node = self.head
        while node != None :
            print (node.data, end=" => ")
            node = node.next


if __name__ == "__main__" :
    s = linked_list()
    s.addToFront(5)
    s.deleteNode(5)
    s.addToFront(8)
    s.addAtEnd("end")
    print (s.getLastNode())
    s.printList()

    print("\n")

#A more complex implementation
class Node :
    def __init__(self, val) :
        self.data = val
        self.next = None

    def getData(self) :
        return self.data

    def getNext(self) :
        return self.next
    
    def setData(self, val):
        self.data = val
    
    def setNext(self, val):
        self.next = val

class linkedList :
    def __init__(self) :
        self.head = None

    def isEmpty(self) :
        """Check if List is Empty """
        return self.head is None

    def add(self, item) :
        """Add Item to list """
        new_node = Node(item)
        new_node.setNext(self.head)
        self.head = new_node

    def size(self) :
        """Return Length / Size of the list """
        count = 0
        current = self.head
        while current is not None:
            count += 1
            current = current.getNext()
        return count

    def search(self, item) :
        """Search if item is in list """
        current =  self.head
        found = False
        while current is not None and not found :
            #will try en optimize later
            if current.getData() is item :
                found = True
            else :
                current = current.getNext()
        return found

    def remove(self, item) :
        """REmove item in list; If not found Raise ValueError """
        current = self.head
        previous = None
        found = False
        while current is not None and not found :
            if current.getData() is item :
                found = True
            else :
                previous = current
                current = current.getNext()
            if found:
                if previous is None :
                    self.head = current.getNext()
                else :
                    previous.setNext(current.getNext())
            else :
                raise ValueError
                print ("value Not Found")

    def insert(self, position, item) :
        """Insert item in specified position and if position is out of bounds, raise IndexError"""
        if position > self.size() - 1 :
            raise IndexError
            print ("Index Out of Bound")
        current = self.head
        previous = None
        pos = 0
        if position is 0 :
            self.add(item)
        else :
            new_node = Node(item)
            while pos < position :
                pos += 1
                previous = current
                current = current.getNext()
            previous.setNext(new_node)
            new_node.setNext(current)

    def index(self, item) :
        """Return Index of item, if not found return None"""
        current = self.head
        pos = 0
        found = False
        while current is not None and not found :
            if current.getData() is item :
                found = True
            else :
                current = current.getNext()
                pos += 1

        if found:
            pass
        else :
            pos = None
        return pos
	
    def pop(self, position = None) :
        """
            If no argument is provided, return and remove the item at the head
            If pos is provided, return and remove the item at the position
            If index is out of bounds raise IndexError
        """
        if position > self.size() :
            raise IndexError
            print ("Index Not found")
        
        current = self.head
        if position is None :
            ret = current.getData()
            self.head = current.getNext()
        else :
            pos = 0
            previous = None
            while pos < position :
                previous = current
                current = current.getNext()
                pos += 1
                ret = current.getData()
            previous.setNext(current.getNext())
        print (ret)
        return ret

    def append(self, item) :
        """Append item to end of list """
        current = self.head
        previous = None
        pos = 0
        length = self.size()
        while pos < length :
            previous = current
            current = current.getNext()
            pos += 1

        new_node = Node(item)
        if previous is None :
            new_node.setNext(current)
            self.head = new_node
        else :
            previous.setNext(new_node)

    def printList(self) :
        current = self.head
        while current is not None:
            print (current.getData())
            current = current.getNext()

if __name__ == "__main__" :
    listExamp = linkedList()
    listExamp.add('l')
    listExamp.add('H')
    listExamp.insert(1, 'e')
    listExamp.append('l')
    listExamp.append('o')
    print(listExamp.index("H"))
    listExamp.printList()

File: linkedList.js
//Simpler Implementation
class node {
    constructor (data = null, next = null) {
        this.data = data;
        this.next = next;
    }
}

class linked_list {
    constructor () {
        this.head = null;
    }

    isEmpty () {
        return this.head === null;
    }

    addToFront (data) {
        this.head = new node(data, this.head);
    }

    addAtEnd (val) {
        if (this.head !== null) {
            this.head = new node(data = val);
        }
        let curr = this.head
        while (curr.next) {
            curr = curr.next;
        }
        curr.next = node(data = val);
    }

    deleteNode (key) {
        let curr = this.head;
        let prev = null;
        while (curr && curr.data !== key) {
            prev = curr;
            curr = curr.next;
        }
        if (prev === null) {
            this.head = curr.next;
        } else if (curr) {
            prev.next = curr.next;
            curr.next = null;
        }
    }

    printList () {
        let node =  this.head;
        while (node !== null) {
            //list += node.data
            console.log(node.data + " => ");
            node = node.next;
        }
    }
}

const test = new linked_list();
test.addToFront(5);
test.addToFront(8);
test.addToFront(9);
test.printList()
console.log("\n")

//A more complex implementation
class Node {
    constructor(val) {
        this.data = val;
        this.next = null;
    }
    getData() {
        return this.data;
    }
    getNext() {
        return this.next;
    }
    setData(val) {
        this.data = val;
    }
    setNext(val) {
        this.next = val;
    }
}

class linkedList {
    constructor () {
        this.head = null;
    }

    isEmpty () {
        return self.head === null;
    }

    add (item) {
        let new_node = new Node(item);
        new_node.setNext(this.head)
        this.head = new_node
    }

    size () {
        let count = 0,
        current = this.head;
        while (current !== null) {
            count += 1;
            current = current.getNext()
        }
        return count;
    }

    search (item) {
        let current = this.head,
        found = false;
        while (current !== null && !found) {
            if (current.getData() === item) {
                found = true;
            } else {
                current = current.getNext();
            }
        }
        return found;
    }

    delete (item) {
        let current = this.head,
        previous = null,
        found = false;
        while (current !== null && !found) {
            if (current.getData() === item) {
                found = true;
            } else {
                previous = current;
                current = current.getNext();
            }
            if (found) {
                if (previous === null) {
                    this.head = current.getNext();
                } else {
                    previous.setNext(current.getNext())
                }
            } else {
                throw ValueError;
                console.log("Value Not Found")
            }
        }
    }

    insert (position, item) {
        if (position > (this.size() - 1)) {
            console.error("Index Out Of Bounds");
        }
        let current = this.head,
        previous = null,
        pos = 0;
        if (position === 0) {
            this.add(item);
        } else {
            let new_node = new Node(item);
            while (pos < position) {
                pos += 1;
                previous = current;
                current = current.getNext();
            previous.setNext(new_node);
            new_node.setNext(current);
            }
        }
    }

    indexOf (item) {
        let current = this.head,
        pos = 0,
        found = false;
        while (current !== null && !found) {
            if (current.getData === item) {
                found = true;
            } else {
                current = current.getNext();
                pos += 1;
            }

            if (!found) {
                pos = null;
            }
            return pos;
        }
    }

    pop (position = null) {
        if (position > this.size()) {
            console.error("Index Not Found");
        }
        let current = this.head;
        if (position === null) {
            let ret = current.getData();
            this.head = current.getNext();
        } else {
            let pos = 0,
            previous = null;
            while (pos < position) {
                previous = current;
                current = current.getNext();
                pos += 1;
                ret = current.getData();
            }
            previous.setNext(current.getData());
            console.log(ret);
            return ret;
        }
    }

    append (item) {
        let current = this.head,
        previous = null,
        pos = 0,
        length = this.size();
        while (pos < length) {
            previous = current;
            current = current.getNext();
            pos += 1;
        }
        let new_node = new Node(item);
        if (previous === null) {
            new_node.setNext(current)
            this.head = new_node
        } else {
            previous.setNext(new_node)
        }
    }

    printList () {
        let current = this.head;
        while (current !== null) {
            console.log(current.getData())
            current = current.getNext();
        }
    }
}


const listExamp = new linkedList()
listExamp.add('l')
listExamp.add('H')
listExamp.insert(1, 'e')
listExamp.append('l')
listExamp.append('o')
listExamp.printList()